<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Алексей! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
    <p class="mb-0"><strong>Комментарий:</strong></p>
<!--     <p class="mb-0"><strong>Исправление:</strong></p> -->
    <p class="mb-0">Здравствуй, Дмитрий, спасибо за проверку. 🔆</p>
</div>

# Учебный Проект → «Викишоп» 📚

>Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 
Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 01</span>
    <p class="mb-0"><strong>Комментарий:</strong></p>
<!--     <p class="mb-0"><strong>Исправление:</strong></p> -->
    <p class="mb-0">Здравствуй, уважаемый ревьювер.</p>
    <p class="mb-0">Делал проект с <code>BERT</code>, но времени не хватило. Сначала импорты <code>transformers</code> и <code>tensorflow</code> раняли <code>kernel</code>. Кстати в тетрадке на kaggle тоже самое. В <code>Google Colab</code> завелись и <code>transformers</code> и <code>tensorflow</code>. Но даже с <code>batch</code> вычисления съели всю оперативку и тетрадка упала..</p>
    <p class="mb-0">Нашел библиотеку <code>pytorch_pretrained_bert</code>. В ней, хоть и криво, но работает. Криво потомучто, например у <code>BertTokenizer</code> нет метода <code>encode</code>. И чтобы хоть както токенизировать нужно пройтись двумя методами <code>tokenize</code> и <code>convert_tokens_to_ids</code>. Но теряются метки начала и конца предложения, не нашел отдельный метод для них в классе.. И тут, казалось бы, можно хотябы поробовать без меток, но <code>cuda</code> не нашлась или не понял как на ubuntu прикрутить <code>cuda</code>. В итоге время на вычесление <code>embeddings</code> какоето немыслимое. Попробовал <code>downsampling</code> чтобы хоть как+то добыть <code>embeddings</code>. Но даже на выборке размером 31814 время вычесления 43+ часа. Уже не успеваю, ведь ещё модель нужно будет обучить..</p>
<p class="mb-0">Вообщем, делаю по второму варианту. Потуги с <code>BERT</code> в окончании проекта. </p>
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>

Здесь сложно выделить какую-то одну ошибку. Но на что я советую обратить внимание:

 - импорт tensorflow. Эта библиотека тебе точно не нужна, используемые нами модели ориентируются на pytorch. 
 - pytorch_pretrained_bert - эта библиотека не поддерживается с 2019 года, и мне кажется что учитывая скорость развития DL библиотек и фреймворков безнадежно устарела. 
    
Решиться на работу с полной базой без GPU это очень смело... Даже если использовать даунсемплинг. Поэтому или работать с очень небольшой выборкой, либо использовать GPU.
    
В Google Colab все очень хорошо работает и при наличии CUDA векторизация всех текстов занимает около 100 минут. Как еще можно ускорить обработку:
    
 - Уменьшить длину последовательности. Сложность работы растет квадратично с длинной последовательности и при длинне 256 токенов должна уменьшиться в 4 раза. Но фактически составляет 37 минут.
 -  Использовать более "легкие" модели, например distilbert. Время работы уменьшаетя до 16 минут на весь датасет.
    

Но с моей стороны будет нечестно не предупредить тебя еще об одной "засаде". На признаках (векторах) полученных с помощью обычных моделей BERT (или дистилированного BERT), обученных на википедии и книжках сложно обучить модели,хорошо детектящие токсичность. Безусловно, этот подход работает и если взять весь датасет то скорее всего получится преодолеть планку 0,7 или может быть даже приблизиться к 0,75. Но нет гарантии что удастся уверенно побить эту планку. 
    
Было бы неплохо использовать чуть более качественные векторные представления. И получить их можно из моделей, уже видевших задачу токсичности.  Можно конечно дообучить BERT на детекцию токсичности и это не очень сложно, однако это заведомо выходит за рамки проекта. БОлее простым решением будет ипользовать веса модели, которая задачу токсичности уже решала. Я могу рекомендовать "unitary/toxic-bert". Кстати эту и другие модели ты можешь найти на HuggingFace по запросу toxic.  
    
    
Кстати,вот ссылка на гугл-колаб, демонстрирующая что все замечательно "взлетает" (я проводил эксперименты) 
    
    
https://drive.google.com/file/d/1nIurdQ8VJDWUKDEO1GcDq-f04ugTVhDt/view?usp=sharing    
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Прошу тебя навести порядок в проекте, удалить лишние ячейки с неудачными экспериментами или просмотрами состояний переменных. </div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 01</span>
    <p class="mb-0"><strong>Комментарий:</strong></p>
<!--     <p class="mb-0"><strong>Исправление:</strong></p> -->
    <p class="mb-0">Почистил. </p>
</div>

Импортируем библиотеки 🎒

In [105]:
import re
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import torch
import nltk
import string

nltk.download('wordnet')
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.corpus import stopwords

from tqdm.notebook import tqdm
from tqdm.notebook import trange
tqdm.pandas()

from numpy.random import RandomState

from matplotlib import rcParams

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import Normalizer

from sklearn.metrics import f1_score
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import make_pipeline

from scipy.sparse import hstack

from catboost import CatBoostClassifier

[nltk_data] Downloading package wordnet to /home/bionic/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /home/bionic/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/bionic/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

In [60]:
# set max columns to none
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 200)

Константы.

In [61]:
RANDOM_SEED = 108108108
RANDOM_NP = RandomState(128)
STOPWORDS = stopwords.words('english')

Функции.

In [62]:
def get_wordnet_pos(word):
    """Map POS tag to first character lemmatize() accepts"""
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {
            "J": wordnet.ADJ,
            "N": wordnet.NOUN,
            "V": wordnet.VERB,
            "R": wordnet.ADV
        }
    return tag_dict.get(tag, wordnet.NOUN)

Подсмотрел на https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 </div>

<hr>

## Загрузим данные

Столбец `text` в нём содержит текст комментария, а `toxic` — целевой признак.

In [63]:
try:
    comments_df = pd.read_csv(
        './datasets/toxic_comments.csv',
        usecols=['text', 'toxic']
    )
    
except FileNotFoundError:
    comments_df = pd.read_csv(
        'https://code.s3.yandex.net/datasets/toxic_comments.csv',
        usecols=['text', 'toxic']
    )
    print('FYI datasets loaded via url')

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Загрузка из облака это беспроигрышный вариант ))</div>

In [64]:
display(
    comments_df.head(2),
    comments_df.sample(2),
    comments_df.tail(2),
    comments_df['toxic'].describe(),
)

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0

text  toxic
81791   "\n\n Systemic risk \n\nI have a couple of sug...      0
104170  "\n\nImage Tagging Image:Aftab2.jpg\n\n This m...      0

text  toxic
159290  And it looks like it was actually you who put ...      0
159291  "\nAnd ... I really don't think you understand...      0

count    159292.000000
mean          0.101612
std           0.302139
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max           1.000000
Name: toxic, dtype: float64

In [65]:
print(
    'Токсичных комментариев: {:.2%}'.format(
            len(comments_df.loc[comments_df['toxic'] == 1].index)
            / len(comments_df.index)
        )
)

Токсичных комментариев: 10.16%


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, исследован баланс классов. Это важная информация для задачи классификации.</div>

Дисбаланс, однако.

Хотел дополнить, но похоже нашел тот же набор данных на kaggle

## Подготовим данные

### Приведём в нижний регистр и почистим твиты

In [66]:
comments_df['text'] = comments_df['text'].str.lower()

Регулярное выражение для ссылок

In [67]:
re_url = re.compile(r'http\S+')

Регулярное выражение электропочт

In [68]:
re_email = re.compile(r'\S*@\S*\s?')

Регулярное выражение для символов

In [69]:
re_symbols = re.compile('[(){}\[\]"\'~:;><»=|—+-_#0123456789!&?$%^*]')

Регулярное выражение для сисвола переноса строк

In [70]:
re_new_line = re.compile('(\n)+')

Регулярное выражение для пробелов

In [71]:
re_space = re.compile(r'(\s)+')

In [72]:
comments_df['text'] = comments_df['text'].apply(
        lambda twt: re.sub(re_url, '', twt)
    )

In [73]:
comments_df['text'] = comments_df['text'].apply(
        lambda twt: re.sub(re_email, '', twt)
    )

In [74]:
comments_df['text'] = comments_df['text'].apply(
        lambda twt: re_symbols.sub(' ', twt)
    )

In [75]:
comments_df['text'] = comments_df['text'].apply(
        lambda twt: re.sub(re_new_line, ' ', twt)
    )

In [76]:
comments_df['text'] = comments_df['text'].apply(
        lambda twt: re.sub(re_space, ' ', twt)
    )

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Чистка выполнена. Отличная тренировка использованию регулярных выражений.</div>

Lemmatization

In [77]:
lemmatizer = WordNetLemmatizer()

In [78]:
def sentence_lemmatizer(sentence):
    return ' '.join(
        [
            lemmatizer.lemmatize(
                w, get_wordnet_pos(w)
            ) for w in nltk.word_tokenize(sentence) if w not in string.punctuation and w not in (STOPWORDS)
        ]
    )

In [82]:
comments_df['text'] = comments_df['text'].progress_apply(
        lambda twt: sentence_lemmatizer(twt)
    )

  0%|          | 0/159292 [00:00<?, ?it/s]

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Лемматизация выполнена абсолютно верно.</div>

In [83]:
comments_df.sample(5)

text  toxic
70006   addition onih thanks cleaning edit prompt add ...      0
100572  visible article one good rank university inclu...      0
95038   reliable source photo annual eminem calender s...      0
84572   thank support thank support candidacy administ...      0
64668   bias curious would seem u single complete page...      0

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Рекомендую собрать все импоры в начале ноутбука.</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Сделал импорты в начале.</p>
</div>

In [84]:
features_train, features_test, target_train, target_test = train_test_split(
        comments_df['text'].to_numpy(),
        comments_df['toxic'].to_numpy(),
        test_size=0.1,
        random_state=RANDOM_SEED
    )

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Векторайзер должен обучаться на обучающем наборе и применяться ко всем остальным.

</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Делим на <code>train</code> <code>test</code> потом <code>Vectorizer</code>.</p>
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    
Данные разделены на выборки. Но я бы посоветовал тебе изменить пропорции и выделить на тест 10%. Причина простая - модели, которые мы обучаем чувствительны к объему обучающих данных. Чем больше слов они увидят в процессе обучения, и оценят их вклад в "токсичность", тем лучше будут модели. А для корректного тестирования и 10% данных вполне достаточно, учитывая немаленький размер датасета.

</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0"><code>test</code> 10%</p>
</div>

In [85]:
tf_idf_vectorizer = TfidfVectorizer()

In [86]:
tf_idf_vect_fitted = tf_idf_vectorizer.fit(features_train)

features_test = tf_idf_vect_fitted.transform(features_test)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
👍 А здесь все совершенно верно,  векторайзер вначале обучен на обучающем наборе и затем применен к тестовому</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Честно говоря я не вижу большой пользы от объединения признаков... Особенно если учесть, что длина каждого признака около 150_000.</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Вот и я не заметил. Ячейки <code>raw</code> удалил.</p>
</div>

In [87]:
del comments_df

Конечно, здесь бы `GridSearchCV` запустить, но для ускорения просто объявим параметры.  
С `GPU` заводится, но памяти не хватает и Jupyter падает..

In [88]:
cat_model = CatBoostClassifier(
        iterations=1250,
        eval_metric='F1',
        # task_type="GPU",
        allow_writing_files=True,
        verbose=250,
    )

In [91]:
cat_model.fit(
    tf_idf_vect_fitted.transform(features_train),
    target_train,
    verbose=False
)

In [92]:
cat_tf_idf_pipe = make_pipeline(
    tf_idf_vectorizer,
    cat_model,
    verbose=True,
)

In [93]:
folds = KFold(
    n_splits=5,
    shuffle=True,
    random_state=RANDOM_SEED
)

In [94]:
f1_cat_cv_score = cross_val_score(
    cat_tf_idf_pipe,
    features_train,
    target_train,
    scoring='f1',
    cv=folds
)

[Pipeline] ... (step 1 of 2) Processing tfidfvectorizer, total=   6.0s
Learning rate set to 0.063604
0:	learn: 0.4995620	total: 1.78s	remaining: 37m 6s
250:	learn: 0.7011030	total: 8m 2s	remaining: 32m 1s
500:	learn: 0.7520329	total: 16m 30s	remaining: 24m 40s
750:	learn: 0.7749748	total: 23m 47s	remaining: 15m 48s
1000:	learn: 0.7926125	total: 29m 53s	remaining: 7m 26s
1249:	learn: 0.8056568	total: 38m 16s	remaining: 0us
[Pipeline]  (step 2 of 2) Processing catboostclassifier, total=38.8min
[Pipeline] ... (step 1 of 2) Processing tfidfvectorizer, total=   7.8s
Learning rate set to 0.063604
0:	learn: 0.4593171	total: 2.65s	remaining: 55m 11s
250:	learn: 0.7010797	total: 8m 35s	remaining: 34m 11s
500:	learn: 0.7531509	total: 17m 8s	remaining: 25m 37s
750:	learn: 0.7746940	total: 24m 17s	remaining: 16m 8s
1000:	learn: 0.7901628	total: 32m 26s	remaining: 8m 4s
1249:	learn: 0.8036287	total: 40m 54s	remaining: 0us
[Pipeline]  (step 2 of 2) Processing catboostclassifier, total=41.5min
[Pipel

In [95]:
print(np.median(f1_cat_cv_score))

0.7644479872459147


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Использование кросс-валидации поверх признаков, полученных с помощью TF-IDF векторайзера порождает утечки.  Как мы получали признаки? Вначале делили на выборки, затем обучали векторайзер только на обучающей выборке и применяли к тестовой. Что происходит при кросс-валидации? Ты разбиваешь большую обучающую выборку на несколько фолдов, из них выделяешь часть на обучение, часть на валидацию.  Но признаки (TF-IDF) уже обучены на всем наборе (обучающий + валидационный). Получается утечка данных. Чтобы избежать её нужно использовать pipeline из векторайзера и модели.
    
Пример такого пайплайна можно посмотреть в официальной документации
    
  https://scikit-learn.org/stable/auto_examples/model_selection/plot_grid_search_text_feature_extraction.html
    
И еще полезные материалы:
    
  https://newtechaudit.ru/pipeline-v-rabote-s-dannymi/  
</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Добавил <code>Pipeline</code> для <code>cross_val_score</code></p>
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Кстати, catboost умеет работать с исходными текстами, а не только с векторным представлением (сам извлекает признаки из текста). Нужно просто передать ей сырой текст и сообщить, в какой колонке он находится. Подробнее можешь посмотреть в официальной документации или вот в этом туториале 
    
https://github.com/catboost/tutorials/blob/master/text_features/text_features_in_catboost.ipynb

Вот еще интересная статья (https://habr.com/ru/post/555064/ )
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Не стоит раньше времени подглядывать в тестовую выборку. Вначале обучим все модели и выберем лучшую по оценкам на крос-валидации и только потом протестируем лучшую модель.</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Сделал тест для лучшей модели.</p>
</div>

In [96]:
log_reg_model = LogisticRegression(
        solver='sag',
        C=15,
        max_iter=400,
        class_weight='balanced',
        random_state=RANDOM_SEED,
    )

In [106]:
log_reg_norm = Normalizer()

In [125]:
log_reg_norm_fitted = log_reg_norm.fit(
        tf_idf_vect_fitted.transform(features_train)
    )

In [116]:
%%time

log_reg_model_fitted = log_reg_model.fit(
    log_reg_norm_fitted.transform(
        tf_idf_vect_fitted.transform(features_train)
    ),
    target_train,
)

CPU times: user 1min 48s, sys: 238 ms, total: 1min 48s
Wall time: 1min 49s


/home/bionic/.local/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


In [110]:
log_reg_tf_idf_pipe = make_pipeline(
    tf_idf_vectorizer,
    log_reg_norm,
    log_reg_model,
    verbose=True,
)

In [111]:
f1_log_reg_cv_score = cross_val_score(
    log_reg_tf_idf_pipe,
    features_train,
    target_train,
    scoring='f1',
    cv=folds
)

[Pipeline] ... (step 1 of 3) Processing tfidfvectorizer, total=   7.7s
[Pipeline] ........ (step 2 of 3) Processing normalizer, total=   0.0s


/home/bionic/.local/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[Pipeline]  (step 3 of 3) Processing logisticregression, total= 1.3min
[Pipeline] ... (step 1 of 3) Processing tfidfvectorizer, total=   7.3s
[Pipeline] ........ (step 2 of 3) Processing normalizer, total=   0.0s


/home/bionic/.local/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[Pipeline]  (step 3 of 3) Processing logisticregression, total= 1.3min
[Pipeline] ... (step 1 of 3) Processing tfidfvectorizer, total=   7.3s
[Pipeline] ........ (step 2 of 3) Processing normalizer, total=   0.0s


/home/bionic/.local/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[Pipeline]  (step 3 of 3) Processing logisticregression, total= 1.2min
[Pipeline] ... (step 1 of 3) Processing tfidfvectorizer, total=   7.5s
[Pipeline] ........ (step 2 of 3) Processing normalizer, total=   0.0s


/home/bionic/.local/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[Pipeline]  (step 3 of 3) Processing logisticregression, total= 1.3min
[Pipeline] ... (step 1 of 3) Processing tfidfvectorizer, total=   7.6s
[Pipeline] ........ (step 2 of 3) Processing normalizer, total=   0.0s


/home/bionic/.local/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[Pipeline]  (step 3 of 3) Processing logisticregression, total= 1.3min


In [113]:
display(np.median(f1_log_reg_cv_score))

0.762798092209857

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Совет: линейные модели очень хорошо работают с признаками, полученными из текстов. С помощью TF-IDF мы получили очень длинные разряженные вектора. И действительно, они очень длинные (длина около 140 000), при этом очень мало значений отличаются от нуля. Так вот, линейные модели гораздо лучше деревьев справляются с такими признаками. Но есть один секрет. Нужно  немного ослабить регуляризацию у логистической регрессии, и сделать это можно подобрав гиперпараметр С (чем он больше тем слабее регуляризация). Оптимальное значение стоит поискать в диапазоне 5-15. </div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
    <p class="mb-0"><strong>Комментарий:</strong></p>
<!--     <p class="mb-0"><strong>Исправление:</strong></p> -->
    <p class="mb-0">Спасибо, класс, попробую.</p>
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Здесь тоже возникают утечки, вязанные с кросс-валидацией над признаками, полученными применением TfidfVectorizer ко всей обучающей выборке.      
</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Добавил <code>Pipeline</code> для <code>cross_val_score</code></p>
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Ниже возникают предупреждения о том, что выбранный солвер не может сойтись на указаном количестве итераций. Попобуй подобрать пару солвер/количество итераций что-бы модель нашла свой глобальный минимум. Попробуй saga или sag.   
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Не стоит раньше времени подглядывать в тестовую выборку. Вначале обучим все модели и выберем лучшую по оценкам на крос-валидации и только потом протестируем лучшую модель.</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Сделал тест для лучшей модели.</p>
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Ты исследовал две модели, давай выберем лучшую используя оценки полученые с помощью кросс-валидации и протестируем её.</div>

<div class="alert alert-info" role="alert" style="border-radius: 7px">
    <span style="font-size:0.7rem">Student's note 02</span>
<!--     <p class="mb-0"><strong>Комментарий:</strong></p> -->
    <p class="mb-0"><strong>Исправление:</strong></p>
    <p class="mb-0">Сделал тест для лучшей модели.</p>
</div>

## Протесстируем `CatBoostClassifier`

In [122]:
cat_pred = cat_model.predict(features_test)

In [123]:
cat_score = f1_score(target_test, cat_pred)

In [124]:
display(
    cat_score
)

0.7640287769784172

## Итог

`CatBoostClassifier` показала чуть лучший результат `fi-score` = 0.764  
`LogisticRegression` всё ёще ругается, дожму её `GridSearch`

Возможно, если запустить поиск гипперпараметров с `GridSearchCV` то можно улучшить `f1`, но долго, уже не успеваю..  
На стандартных параметрах получилось перевалить едва за значение `f1` 0.75..

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>
Давай подведем итоги. В целом с проектом ты справился - текст предобработан, извлечены признаки и обучены классификаторы. Достигнуто требуемое значение метрики f1.
    
Молодец, что корректно произвел чистку и лемматизацию. 
    
Свои соображения по поводу BERT я тебе очень подробно расписал в первом комментарии.
    
Очень прошу тебя привести проект в порядок. Сейчас создается ощущение, что ты очень торопился, быстро получил рабочее решение, но навести порядок в проекте не успел.
    
Что нужно дорабоать  обязательно (красные комментарии).
    
 - Векторайзеры должны обучаться на обучающей выборке, а не на всех даных.    
 - Нельзя раньше времени подглядывать в тестовую выборку. Вначале выбираем лучшую модель по кросс-валидации и только потом делаем тестирование лучшей модели.
 
Доработок немного, поэтому уверен, что ты со всем справишься.     
    
Жду твоих исправлений :)
</div>

<hr>

>Все ли шаги по инструкции выполняете?  
Как готовите данные?  
Как анализируете данные?  
Какие модели и гиперпараметры рассматриваете?  
Не дублируете ли код?  
Какие выводы делаете?  
Следите ли за структурой проекта?  
Поддерживаете аккуратность кода?  

$$ \large
TF = \dfrac{t}{n}
$$

>где t (от англ. term) — количество употребления слова, а n — общее число слов в тексте.  
IDF нужна в формуле, чтобы уменьшить вес слов, наиболее распространённых в любом другом тексте заданного корпуса. IDF зависит от общего числа текстов в корпусе (D) и количества текстов, в которых это слово встречается (d). 

$$ \large
IDF = \log_{10}\dfrac{D}{d}
$$

$$ \large
TFIDF = TF * IDF
$$

> Большая величина TF-IDF говорит об уникальности слова в тексте по отношению к корпусу.  
Чем чаще оно встречается в конкретном тексте и реже в остальных, тем выше значение TF-IDF.